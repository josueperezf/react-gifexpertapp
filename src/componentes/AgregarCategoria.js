import React, { useState } from 'react'
import PropTypes from 'prop-types';


export const AgregarCategoria = ({setcategorias}) => {
    const [inputValue, setinputValue] = useState('');

    const handleInputChange = (event)=>{
        setinputValue(event.target.value);
    }

    const handleSubmit = (e)=>{
        e.preventDefault();
        // const valor = e.target.elements[0].value;
        if(inputValue.trim().length > 0) {
            setcategorias((categorias)=>[inputValue, ...categorias]);
            setinputValue('');
        }
    }
    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value = {inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

AgregarCategoria.propTypes = {
    setcategorias:PropTypes.func.isRequired
}