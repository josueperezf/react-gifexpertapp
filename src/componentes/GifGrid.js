import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import GifGridItem from './GifGridItem';
import PropTypes from 'prop-types'

const GifGrid = ({categoria}) => {
    const {data:imagenes, loading} = useFetchGifs(categoria);
    /*
    // el siguiente hook hace evaita que cada vez que se realice un cambio en el html realice la llamada http
    useEffect (() => {
        getGifs(categoria).then((imgs)=>{
            setimages(imgs);
        });
    },[categoria])
    */
    // en la anterior linea, [categoria] se puede quitar, pero fernando la coloca y dice que es para que si en algun momento cambia el valor de la categoria, entonces todo lo que esta dentro de 'useEffect' se ejecutaria nuevamente
    
    return (
        <>
        <h3 className=" animate__animated animate__fadeIn">{categoria}</h3>
        {/** la siguiente linea es como un operador ternario, pero como no quiere evaluar el de lo contrario, se unsa el && */}
        { loading && <p className="animate__animated animate__flash" > Cargando</p> }
        <div className="card-grid">
            { /**  
             * en la seccion GifGridItem se le va a enviar el img, pero fernando hacer una destructuracion {...img}
             * con ello logra que GifGridItem no reciba asi ({img}), sino directamente ({id, title, utl})
            */}
                {
                    
                    imagenes.map(img=>{
                        return <GifGridItem key={img.id}  {...img} />
                    })
                    
                }
        </div>
        </>
    )
}

GifGrid.propTypes = {
    categoria: PropTypes.string.isRequired
}

export default GifGrid
