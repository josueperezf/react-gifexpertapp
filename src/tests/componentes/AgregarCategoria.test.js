import { shallow } from "enzyme";
import { AgregarCategoria } from "../../componentes/AgregarCategoria";

describe('Pruebas en el componente <AgregarCategoria/>',()=>{

    //const setcategorias = ()=>{};
    const setcategorias = jest.fn();
    let wrapper =  shallow(<AgregarCategoria setcategorias={setcategorias} />);

    beforeEach(()=>{
        // aqui coloco todo lo que quiero que se reinicialice
        jest.clearAllMocks();
        wrapper =  shallow(<AgregarCategoria setcategorias={setcategorias} />);
    });

    test('debe mostrarse correctamente', () => {
        // la siguiente linea es para que haga el Snapshot del html del componente
        expect(wrapper).toMatchSnapshot();
    });
    
    test('debe de cambiar la caja de texto', () => {
        const input =  wrapper.find('input');
        const value = 'hola a todos';
        // en la siguiente linea, el evento change, pasa un evento, para hacerlo, le pasamos un objeto vacio
        // el target value, es por que el componente hace esto, setinputValue(event.target.value); con las llaves le mandamos el event, pero necesita el objeto target y la propiedad value
        input.simulate('change',{target:{value}});
    });

    test('NO debe postear la informacion submit', () => {
        // verifica si realmente NO esta enviado data cuando ejecuta el submit y el input no tiene informacion
        // coloco la siguiente linea ya que como en el pasado test se paso data 
        // const wrapper =  shallow(<AgregarCategoria setcategorias={setcategorias} />);
        // el submit emite un evento, para simularlo se pasa un objeto, como el componente que evaluamos tenemos q usaremos el llamado a la funcion preventDefault, entonces hay q enviarla para que no arroje error
        wrapper.find('form').simulate('submit',{ preventDefault: ()=>{} });
        // la siguiente linea usa el setcategorias que es similiar al useState del componente
        // .not es una negacion
        //  toHaveBeenCalled para saber si fue llamado
        // pregunta si el metodo setcategorias no fue llamado
        expect(setcategorias).not.toHaveBeenCalled();
    });
    
    test('debe de llamar al setcategorias y limpiar la caja de texto', () => {
        // simular inputChange
        // simular el submit
        // setcategorias se debe haber llamado
        // el valor del input debe estar como string vacio ''
        wrapper.find('input').simulate('change',{ target:{value:'valor de prueba'}});
        // const valorAsignado = wrapper.find('input').prop('value');
        wrapper.find('form').simulate('submit',{preventDefault:()=>{}});
        // preguntar si la funcion setcategorias fue llamada una vez
        expect(setcategorias).toHaveBeenCalled();
        // la siguiente seccion es para saber si setcategorias fue llamado mediante una funcion javascript
        expect(setcategorias).toHaveBeenCalledWith(expect.any(Function));

        expect('').toBe(wrapper.find('input').prop('value'));
    })
    
});