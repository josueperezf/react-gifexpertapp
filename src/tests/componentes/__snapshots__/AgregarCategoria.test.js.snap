// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AgregarCategoria/> debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
